### PR TITLE
81 supply custom rate constants to fortran

### LINKED ISSUE #81: Supply custom rate constants to fortran
#79 and #80 added support for custom rate constants to be set, but the fortran stuff wasn't fully finished. Finish it.

*Acceptance Criteria*

- the fortran wrapper tests can pass in concentrations and custom rate constant parameters that used during MICM solving without prior knowledge of the ordering of these arrays in the MICM state

### PR DESCRIPTION
**Pull request by K20shores**
Closes #81

### CODE REVIEW

File: fortran/micm_core.F90 [hunk 1]
```diff
@@ -3,12 +3,13 @@ module micm_core
    use iso_c_binding, only: c_ptr, c_char, c_int, c_double, c_null_char, c_size_t, c_f_pointer
    implicit none
 
-   public :: micm_t
+   public :: micm_t, Mapping
```
Comment (mattldawson) [inline fortran/micm_core.F90:6]: I think we've been using the style `mapping_t` for derived types in Fortran

**Review by mattldawson — CHANGES_REQUESTED** [review]
Looks great! Just one minor suggestion

File: fortran/micm_core.F90 (hunk 1 above)
Comment (K20shores) [reply fortran/micm_core.F90:6]: ah, yes. fixing that now

File: fortran/test/fetch_content_integration/CMakeLists.txt [hunk 2]
```diff
@@ -11,7 +11,7 @@ include(FetchContent)
 
 FetchContent_Declare(musica-fortran
   GIT_REPOSITORY https://github.com/NCAR/musica.git
-  GIT_TAG        79-allow-updates-to-custom-rate-constant-parameters-and-mapping-by-name
+  GIT_TAG        81-supply-custom-rate-constants-to-fortran
```
[the remark this answers is not in the document]
Comment (K20shores) [reply fortran/test/fetch_content_integration/CMakeLists.txt:14]: yes, though that will cause the tests to fail for now

File: fortran/test/fetch_content_integration/test_micm_fort_api.F90 [hunk 3]
```diff
@@ -1,28 +1,42 @@
...
-  integer(c_int)                :: num_concentrations
-  real(c_double), dimension(5)  :: concentrations
-  integer                       :: errcode
-  character(len=7)              :: config_path
+  integer(c_int)                :: num_concentrations, num_user_defined_reaction_rates
+  real(c_double), dimension(5)  :: concentrations 
+  real(c_double), dimension(3)  :: user_defined_reaction_rates 
+  integer                       :: errcode, i
+  character(len=256)            :: config_path
+  type(mapping_t)                 :: the_mapping
```
Comment (boulderdaze) [inline fortran/test/fetch_content_integration/test_micm_fort_api.F90:16]: Could you make spacing uniformly maybe?

File: fortran/test/fetch_content_integration/test_micm_fort_api.F90 [hunk 4]
```diff
@@ -1,28 +1,42 @@
...
 
+  do i = 1, micm%species_ordering_length
+    the_mapping = micm%species_ordering(i)
+    print *, "Species Name:", the_mapping%name(:the_mapping%string_length), ", Index:", the_mapping%index
+  end do
+  do i = 1, micm%user_defined_reaction_rates_length
+    the_mapping = micm%user_defined_reaction_rates(i)
+    print *, "User Defined Reaction Rate Name:", the_mapping%name(:the_mapping%string_length), ", Index:", the_mapping%index
+  end do
+
```
Comment (boulderdaze) [inline fortran/test/fetch_content_integration/test_micm_fort_api.F90:39]: I'm curious how we can do this for the CCPP wrapper. Would CAM-SIMA take care of this?

File: fortran/test/unit/test_musica_fortran_interface.F90 [hunk 5]
```diff
@@ -1,28 +1,42 @@
...
+  real(c_double), dimension(5)  :: concentrations 
+  real(c_double), dimension(3)  :: user_defined_reaction_rates 
+  integer                       :: errcode, i
+  character(len=256)            :: config_path
+  type(mapping_t)                 :: the_mapping
 
   time_step = 200
   temperature = 272.5
-  pressure = 101253.3
+  pressure = 101253.4
```
Comment (boulderdaze) [inline fortran/test/unit/test_musica_fortran_interface.F90:20]: haha just curious, what made you change the decimal number?

**Review by boulderdaze — APPROVED** [review]

File: fortran/test/fetch_content_integration/test_micm_fort_api.F90 (hunk 4 above)
Comment (K20shores) [reply fortran/test/fetch_content_integration/test_micm_fort_api.F90:39]: We will have to use these in cam sima to properly order things

File: fortran/test/unit/test_musica_fortran_interface.F90 (hunk 5 above)
Comment (K20shores) [reply fortran/test/unit/test_musica_fortran_interface.F90:20]: I don't have a memory of even doing that

**Pull request merged**